### PR TITLE
Feature/wemo listening server subscription refactor

### PR DESCRIPTION
**Pull request by cjswedes**
The first commit shows the design I was going for before I found a few hub firmware bugs that resulted in a change of design. Unless there is an objection, I think I will leave it there to look back on once the fw bugs are fixed. The second commit contains the working subscription server.

I have done some testing with 5 switch devices (3 different models), and all seem to be maintaining their subscriptions. I will do further testing with migration and the motion devices prior to merging.

### CODE REVIEW

File: drivers/SmartThings/wemo/src/parser.lua [hunk 1]
```diff
@@ -30,49 +30,41 @@ end
 
 local function handle_binary_state(device, value)
   -- parses wemo insight style ("8|1611850428|58|...") state, works just fine for a single value too
```
Comment (dljsjr) [inline drivers/SmartThings/wemo/src/parser.lua:32]: Might be nice if there was a link to a doc that describes what the format being parsed looks like but I'm assuming that document doesn't exist.

File: drivers/SmartThings/wemo/src/parser.lua [hunk 2]
```diff
@@ -30,49 +30,41 @@ end
...
-  log.debug("binary state value=", value)
   local vals = string.gmatch(value, "%d+")
```
Comment (dljsjr) [inline drivers/SmartThings/wemo/src/parser.lua:33]: So is the idea that it's just a bunch of pipe-delimited numbers then?

File: drivers/SmartThings/wemo/src/init.lua [hunk 3]
```diff
@@ -38,40 +39,7 @@ local profiles = {
...
-  if server.listen_sock ~= nil then
-    server.listen_sock:close()
-  end
-  --TODO unregister channel handler? Seems possible to get in a state
-  -- where sockets are never gc'd because they are held by the driver.
-  server = {}
-end
+local server
 
 -- build a exponential backoff time value generator
```
Comment (dljsjr) [inline drivers/SmartThings/wemo/src/init.lua:44]: This function is in `st.utils` now.

File: drivers/SmartThings/wemo/src/init.lua [hunk 4]
```diff
@@ -121,32 +89,39 @@ local function device_init(driver, device)
...
+  --Persist to avoid issues with driver restart
   device:set_field("ip", info.ip, { persist = true })
   device:set_field("port", info.port, { persist = true })
   device:set_field("serial_num", info.serial_num, { persist = true })
 
-  --TODO maybe we should call_on_schedule with the device thread, and the polling.
-  -- instead of doing the resubscribe for all devices at once.
-  protocol.subscribe(server, device)
-end
+  local jitter = 2 * math.random()
```
Comment (dljsjr) [inline drivers/SmartThings/wemo/src/init.lua:110]: What's the variance for?

**Review by dljsjr — APPROVED** [review]
Some style comments/clarifying questions but LGTM otherwise

File: drivers/SmartThings/wemo/src/parser.lua (hunk 1 above)
Comment (cjswedes) [reply drivers/SmartThings/wemo/src/parser.lua:32]: wemo? documentation? I wish. These numbers are still a mystery to me.

File: drivers/SmartThings/wemo/src/parser.lua (hunk 2 above)
Comment (cjswedes) [reply drivers/SmartThings/wemo/src/parser.lua:33]: Yeah, my guess is that they are power/energy numbers from the insight plug, but I have yet to make sense of them, and there's no documentation that I have found.

File: drivers/SmartThings/wemo/src/init.lua (hunk 4 above)
Comment (cjswedes) [reply drivers/SmartThings/wemo/src/init.lua:110]: During Migration, driver startup, and polling intervals, it spreads network traffic out. If 10 wemo devices are added, we would send 10 http request in like a quarter second if all devices respond to the first discovery message. And then every 60 sec we would send 10 more. This is an attempt to spread that out a little bit. The same goes for re subscribing. It makes the biggest difference during driver restart.

Im not sure it solves a problem that exists, but these devices are unreliable and I want to give them the best chance.

File: drivers/SmartThings/wemo/src/init.lua [hunk 5]
```diff
@@ -113,74 +52,85 @@ local function device_init(driver, device)
...
-    log.info_with({ hub_logs = true }, "[" .. device.device_network_id .. "] polling device")
-    protocol.poll(driver, device)
-  end
+    --Sometimes wemos just stop responding to ssdp even though they are connected to the network.
+    --Persist to avoid issues with driver restart
+    device:set_field("ip", info.ip, { persist = true })
+    device:set_field("port", info.port, { persist = true })
+    device:set_field("serial_num", info.serial_num, { persist = true })
+    if driver.server and (ip ~= info.ip or port ~= info.port) then
+      print("not working how expected!!!!!!!!!!1")
```
Comment (sjl2178) [inline drivers/SmartThings/wemo/src/init.lua]: nit: I think 1 and ! are equivalent but not everyone will

**Review by sjl2178 — APPROVED** [review]

**Comment by cjswedes** [comment]
Merging after doing a migration test with 7 switches and 3 motion sensors. There are some pain points in the integration, but nothing as bad as the DTH backed devices and the integration overall works much better in the driver IMO.

**Comment by greens** [comment]
Squash your commits next time, please.

**Comment by cjswedes** [comment]
I intentionally left multiple commits because they represent specific pieces of functionality, and improvements. It aids in review both in the PR and when looking back in git history imo.

**Pull request merged**